graphjet-core/src/main/java/com/twitter/graphjet/algorithms/counting/TopSecondDegreeByCountTweetRecsGenerator.java

```diff
@@ -109,6 +105,33 @@ private static boolean isLessThantMinUserSocialProofSize(
     return !keep;
   }
 
+  private static boolean isLessThanMinUserSocialProofSizeCombined(
+    SmallArrayBasedLongToDoubleMap[] socialProofs,
+    int minUserSocialProofSize,
+    Set<byte[]> socialProofTypeUnions
+  ) {
+    boolean keep = false;
+
+    for (byte[] socialProofTypeUnion: socialProofTypeUnions) {
+      Set<Long> uniqueNodes = new HashSet<Long>();
+      for (byte socialProofType: socialProofTypeUnion) {
```

Comment:
Awesome suggestions. Thanks for the feedback!

1) I will use LongOpenHashSet as opposed to a regular HashSet.
2) I will break out of both for-loops as soon as the size >= minUserSocialProofSize. This means that the LongOpenHashSet for each candidate will be as large as twice as much  the minUserSocialProofSize (worst case due to the doubling of the hashSet).
3) I think I am already checking for that precondition mentioned above in the order of my if statement: 

(a) isLessThanMinUserSocialProofSize(nodeInfo.getSocialProofs(), minUserSocialProofSize) &&

(b) isLessThanMinUserSocialProofSizeCombined(
            nodeInfo.getSocialProofs(), minUserSocialProofSize, request.getSocialProofTypeUnions()
          )

The if condition is done as soon as (a) is false. Which means we only call (b) if all social proofs of that candidate are less than the threshold.